### PR TITLE
Improved gain recalculation

### PR DESCRIPTION
**Pull request by larsgottesbueren**
No more memory overheads!

### CODE REVIEW

File: mt-kahypar/datastructures/hypergraph_common.h [hunk 1]
```diff
@@ -80,6 +80,14 @@ struct Move {
   PartitionID to = -1;
   HypernodeID node = invalidNode;
   Gain gain = invalidGain;
+
+  bool isValid() const {
+    return from != -1;
```
Comment (kittobi1992) [inline mt-kahypar/datastructures/hypergraph_common.h:85]: -1 == kInvalidPartition

File: mt-kahypar/datastructures/partitioned_hypergraph.h [hunk 2]
```diff
@@ -162,10 +162,6 @@ class PartitionedHypergraph {
     return _hg->initialNumEdges();
   }
 
-  HyperedgeID numGraphEdges() const {
-    return _hg->numGraphEdges();
-  }
-
   HyperedgeID numNonGraphEdges() const {
```
Comment (kittobi1992) [inline mt-kahypar/datastructures/partitioned_hypergraph.h:165]: you are lucky that this function is not used throughout the code -> remove

File: mt-kahypar/partition/refinement/fm/global_rollback.h [hunk 3]
```diff
@@ -31,30 +31,19 @@ class GlobalRollback {
...
-      resetStoredMoveIDs();
+          context(context),
+          max_part_weight_scaling(context.refinement.fm.rollback_balance_violation_factor),
+          num_parts(numParts),
+          ets_recalc_data(vec<RecalculationData>(num_parts)),
+          last_recalc_round(),
+          round(1)
+  {
+    if (context.refinement.fm.rollback_sensitive_to_num_moves && context.refinement.fm.rollback_parallel) {
+      last_recalc_round.resize(hg.initialNumEdges(), CAtomic<uint32_t>(0));
```
Comment (kittobi1992) [inline mt-kahypar/partition/refinement/fm/global_rollback.h:42]: You know the rule => Move to Memory Pool

File: mt-kahypar/partition/refinement/fm/global_rollback.h [hunk 4]
```diff
@@ -31,30 +31,19 @@ class GlobalRollback {
   static constexpr bool enable_heavy_assert = false;
 public:
   explicit GlobalRollback(const Hypergraph& hg, const Context& context, PartitionID numParts) :
```
Comment (kittobi1992) [inline mt-kahypar/partition/refinement/fm/global_rollback.h:33]: numParts == context.partition.k?

File: mt-kahypar/partition/refinement/fm/global_rollback.cpp [hunk 5]
```diff
@@ -235,115 +204,94 @@ namespace mt_kahypar {
...
-          PartitionID pv;
-          if (v_moved) {
-            const MoveID local_move_id_of_v = move_id_of_v - firstMoveID;
-            const Move& move_of_v = move_order[local_move_id_of_v];
-            if (local_move_id_of_v < localMoveID) {
-              pv = move_of_v.to;
-            } else {
-              pv = move_of_v.from;
-            }
+      if (num_parts <= static_cast<int>(2 * phg.edgeSize(e))) {
```
Comment (kittobi1992) [inline mt-kahypar/partition/refinement/fm/global_rollback.cpp:256]: some comment what is going on here in this case distinction would be helpful

File: mt-kahypar/partition/refinement/fm/global_rollback.cpp [hunk 6]
```diff
@@ -235,115 +204,94 @@ namespace mt_kahypar {
...
           }
+        }
+      }
+    };
 
-          if (pv == m.to) {
-            gain += edgeWeight;
-          } else if (pv == m.from) {
-            gain -= edgeWeight;
+    if (!context.refinement.fm.rollback_sensitive_to_num_moves) {
```
Comment (kittobi1992) [inline mt-kahypar/partition/refinement/fm/global_rollback.cpp:273]: weird naming: rollback_sensitive_to_num_moves. Maybe some explanation in context.h would be helpful

**Review by kittobi1992 — APPROVED** [review]
some minor comments

File: mt-kahypar/partition/refinement/fm/global_rollback.cpp (hunk 6 above)
Comment (larsgottesbueren) [reply mt-kahypar/partition/refinement/fm/global_rollback.cpp:273]: yup. any alternative suggestions? I thought about calling it iter-moves but that doesn't really explain what it does either

**Pull request merged**